### PR TITLE
resolve: Levenshtein-based suggestions for non-import paths

### PR DESCRIPTION
**Pull request by petrochenkov**
This patch addresses both items from https://github.com/rust-lang/rust/issues/30197#issuecomment-264846000 and therefore implements the largest part of https://github.com/rust-lang/rust/issues/30197.

r? @jseyfried

### CODE REVIEW

File: src/test/ui/resolve/levenshtein.stderr [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+  --> $DIR/levenshtein.rs:19:10
+   |
+19 | type A = Baz; // Misspelled type name.
+   |          ^^^ did you mean `Bar`?
+
+error[E0412]: unresolved type `Opiton`
+  --> $DIR/levenshtein.rs:20:10
+   |
+20 | type B = Opiton<u8>; // Misspelled type name from the prelude.
+   |          ^^^^^^^^^^ did you mean `Option`?
```
Comment (nagisa) [inline src/test/ui/resolve/levenshtein.stderr]: Span seems weird here. Either suggest `Option<u8>` or have span to only point to `Opiton`.

Comment (petrochenkov) [inline src/test/ui/resolve/levenshtein.stderr]: >have span to only point to `Opiton`

That would be nice, but path segments and identifiers in them don't have their own spans.

>suggest `Option<u8>`

In fact, type arguments are intentionally dropped from paths in resolve error messages - they are usually irrelevant and mostly bring noise. I tried to include them and it looked worse. I can make a separate branch with this change so you can look and compare.

Comment (nagisa) [inline src/test/ui/resolve/levenshtein.stderr]: > I tried to include them and it looked worse. I can make a separate branch with this change so you can look and compare.

Never mind then.

**Comment by jseyfried** [comment]
Nice! @bors r+

**Comment by petrochenkov** [comment]
>:broken_heart: Test failed - status-travis

`HashSet` makes suggestions non-deterministic, will fix today.

**Comment by petrochenkov** [comment]
OSX linker segfault
@bors retry

**Comment by alexcrichton** [comment]
@bors: retry

* osx linker segfaulted #38878

**Comment by petrochenkov** [comment]
@bors r=jseyfried
(this needed rebase after https://github.com/rust-lang/rust/pull/38890)

**Pull request merged**